tests/test_xlora.py

```diff
@@ -0,0 +1,172 @@
+# Copyright 2023-present the HuggingFace Inc. team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+
+import torch
+from transformers import AutoModelForCausalLM, AutoTokenizer
+
+from peft import LoraConfig, PeftType, TaskType, XLoraConfig, get_peft_model
+
+
+model_id = "facebook/opt-125m"
+tokenizer = AutoTokenizer.from_pretrained(model_id, trust_remote_code=True, device_map="cuda:0")
+tmp_path = "/tmp/xlora_test"
+os.makedirs(tmp_path, exist_ok=True)
+
+for i in range(1, 9):
+    torch.manual_seed(i)
+    lora_config = LoraConfig(task_type="CAUSAL_LM", init_lora_weights=False)
+    model = AutoModelForCausalLM.from_pretrained(model_id)
+    peft_model = get_peft_model(model, lora_config)
+    peft_model.save_pretrained(f"{tmp_path}/checkpoint-{i}")
+    print(f"finished {i} of 8")
```

Comment:
Great that you added more tests, this should give us more confidence that everything works.

Running this code block on the file root is problematic, however, because it means this is always run, even if we don't need it. Additionally, this might not work on Windows (which uses `\` for paths and has a different temp folder) and it doesn't clean up the temporary file.

However, this is a perfect opportunity to use pytest fixtures! In case you haven't worked with fixtures yet, I have started to code up something that you can try. Delete all this code and instead adjust the test class and the first test like so:

```python
class TestXlora:
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    model_id = "facebook/opt-125m"
    num_loras = 4

    @pytest.fixture(scope="class")
    def tmp_dir(self, tmp_path_factory):
        # create a class-scoped temp directory
        return tmp_path_factory.mktemp("xlora")

    @pytest.fixture(scope="class")
    def saved_lora_adapters(self, tmp_dir):
        file_names = []
        for i in range(1, self.num_loras + 1):
            torch.manual_seed(i)
            lora_config = LoraConfig(task_type="CAUSAL_LM", init_lora_weights=False)
            model = AutoModelForCausalLM.from_pretrained(self.model_id)
            peft_model = get_peft_model(model, lora_config)
            file_name = os.path.join(tmp_dir, f"checkpoint-{i}")
            peft_model.save_pretrained(file_name)
            file_names.append(file_name)
        return file_names

    def test_functional(self, saved_lora_adapters):
        tokenizer = AutoTokenizer.from_pretrained(self.model_id, trust_remote_code=True, device_map=self.device)

        model = AutoModelForCausalLM.from_pretrained(self.model_id)
        model.config.use_cache = False
        adapters = {str(i): file_name for i, file_name in enumerate(saved_lora_adapters)}

        peft_config = XLoraConfig(
            task_type=TaskType.CAUSAL_LM,
            peft_type=PeftType.XLORA,
            hidden_size=model.config.hidden_size,
            xlora_depth=8,
            adapters=adapters,
        )
        model = get_peft_model(model, peft_config).to(self.device)

        model.enable_scalings_logging()
        inputs = tokenizer.encode("Python is a", add_special_tokens=False, return_tensors="pt")
        outputs = model.generate(
            input_ids=inputs.to("cuda"),
            max_new_tokens=32,
        )
        text = tokenizer.batch_decode(outputs[: inputs.shape[1] :].detach().cpu().numpy(), skip_special_tokens=True)
        # TODO: do any check on the text?
```

Note that I made a few more changes, like setting `self.model_id`, ensuring that we always use `self.device` instead of hard-coding `cuda`, and going down to 4 LoRAs to speed things up a little. The other test functions would have to be adjusted the same way as `test_functional` to use the new fixture and attributes.

Don't hesitate to ask if you have questions about this.

Reply:
Wow, that is really amazing! Thank you for letting me know, I have updated the code.